cs4660/graph/graph.py

```diff
@@ -0,0 +1,292 @@
+"""
+
+from io import open
+from operator import itemgetter
+
+def construct_graph_from_file(graph, file_path):
+    """
+    TODO: read content from file_path, then add nodes and edges to graph object
+
+    note that grpah object will be either of AdjacencyList, AdjacencyMatrix or ObjectOriented
+
+    In example, you will need to do something similar to following:
+
+    1. add number of nodes to graph first (first line)
+    2. for each following line (from second line to last line), add them as edge to graph
+    3. return the graph
+    """
+    nodes = []
+    edges = []
+
+    f = open(file_path, encoding='utf-8')
+    text = f.readlines()
+    size = int(text[0])
+    
+    for n in range(size):
+        nodes.append(Node(n))
+
+    for d in range(1, len(text), 1):
+        lines = text[d].split(":")
+        from_node = getNode(int(lines[0]), nodes)
+        to_node = getNode(int(lines[1]), nodes)
+        weight = int(lines[2])
+        edges.append(Edge(from_node, to_node, weight))
+
+    for node in nodes:
+        graph.add_node(node)
+    for edge in edges:
+        graph.add_edge(edge)
+
+    return graph
+
+def getNode(data, nodes):
+    for i in nodes:
+        if i.data == data:
+            return i
+    
+    return None
+
+class Node(object):
+    """Node represents basic unit of graph"""
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+    """Edge represents basic unit of graph connecting between two edges"""
+    def __init__(self, from_node, to_node, weight):
+        self.from_node = from_node
+        self.to_node = to_node
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+
+    def __eq__(self, other_node):
+        return self.from_node == other_node.from_node and self.to_node == other_node.to_node and self.weight == other_node.weight
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+
+    def adjacent(self, node_1, node_2):
+        if node_2 in self.adjacency_list[node_1]:
```

Comment:
data structure used here is a bit off, it doesn't have the capability to store the edge weight. This is okay for homework 1 but you will soon see issue in homework 2 on Dijkstra Search and A* Search. Please fix it for homework 2.

Reply:
It should be a dictionary of node to a list of **edges** instead of dictionary of node to a list of nodes.